diff --git a/source/lexbor/dom/interfaces/node.h b/source/lexbor/dom/interfaces/node.h
index acd0c1c..f436257 100755
--- a/source/lexbor/dom/interfaces/node.h
+++ b/source/lexbor/dom/interfaces/node.h
@@ -46,6 +46,8 @@ struct lxb_dom_node {
     uintptr_t              prefix;     /* lowercase: lalala */
     uintptr_t              ns;         /* namespace */
 
+    size_t                 source_location; /* char offset to the source HTML */
+
     lxb_dom_document_t     *owner_document;
 
     lxb_dom_node_t         *next;
diff --git a/source/lexbor/html/tokenizer.c b/source/lexbor/html/tokenizer.c
index 741bced..6343c65 100755
--- a/source/lexbor/html/tokenizer.c
+++ b/source/lexbor/html/tokenizer.c
@@ -309,6 +309,7 @@ lxb_html_tokenizer_chunk(lxb_html_tokenizer_t *tkz, const lxb_char_t *data,
 
     tkz->is_eof = false;
     tkz->status = LXB_STATUS_OK;
+    tkz->first = data;
     tkz->last = end;
 
     while (data < end) {
diff --git a/source/lexbor/html/tokenizer.h b/source/lexbor/html/tokenizer.h
index ba9602f..08d0d9a 100755
--- a/source/lexbor/html/tokenizer.h
+++ b/source/lexbor/html/tokenizer.h
@@ -72,6 +72,7 @@ struct lxb_html_tokenizer {
     lxb_char_t                       *pos;
     const lxb_char_t                 *end;
     const lxb_char_t                 *begin;
+    const lxb_char_t                 *first;
     const lxb_char_t                 *last;
 
     /* Entities */
diff --git a/source/lexbor/html/tree.c b/source/lexbor/html/tree.c
index 8c42990..6b19c32 100755
--- a/source/lexbor/html/tree.c
+++ b/source/lexbor/html/tree.c
@@ -434,6 +434,8 @@ lxb_html_tree_create_element_for_token(lxb_html_tree_t *tree,
         return NULL;
     }
 
+    node->source_location = token->begin - tree->tkz_ref->first;
+
     lxb_status_t status;
     lxb_dom_element_t *element = lxb_dom_interface_element(node);
 
@@ -484,6 +486,7 @@ lxb_html_tree_append_attributes(lxb_html_tree_t *tree,
 
         attr->node.local_name = token_attr->name->attr_id;
         attr->node.ns = ns;
+        attr->node.source_location = token_attr->name_begin - tree->tkz_ref->first;
 
         /* Fix for adjust MathML/SVG attributes */
         if (tree->before_append_attr != NULL) {
@@ -767,6 +770,8 @@ lxb_html_tree_insert_character_for_data(lxb_html_tree_t *tree,
 
     lxb_dom_interface_text(text)->char_data.data = *str;
 
+    text->source_location = tree->tkz_ref->begin - tree->tkz_ref->first;
+
     if (ret_node != NULL) {
         *ret_node = text;
     }
@@ -812,6 +817,8 @@ lxb_html_tree_insert_comment(lxb_html_tree_t *tree,
         return NULL;
     }
 
+    node->source_location = token->begin - tree->tkz_ref->first;
+
     lxb_html_tree_insert_node(pos, node, ipos);
 
     return comment;
@@ -831,6 +838,8 @@ lxb_html_tree_create_document_type_from_token(lxb_html_tree_t *tree,
         return NULL;
     }
 
+    doctype_node->source_location = token->begin - tree->tkz_ref->first;
+
     doc_type = lxb_dom_interface_document_type(doctype_node);
 
     /* Parse */
