diff --git a/source/lexbor/dom/interfaces/node.h b/source/lexbor/dom/interfaces/node.h
index acd0c1c..f436257 100755
--- a/source/lexbor/dom/interfaces/node.h
+++ b/source/lexbor/dom/interfaces/node.h
@@ -46,6 +46,8 @@ struct lxb_dom_node {
     uintptr_t              prefix;     /* lowercase: lalala */
     uintptr_t              ns;         /* namespace */
 
+    size_t                 source_location; /* char offset to the source HTML */
+
     lxb_dom_document_t     *owner_document;
 
     lxb_dom_node_t         *next;
diff --git a/source/lexbor/html/tokenizer.c b/source/lexbor/html/tokenizer.c
index 741bced..6343c65 100755
--- a/source/lexbor/html/tokenizer.c
+++ b/source/lexbor/html/tokenizer.c
@@ -309,6 +309,7 @@ lxb_html_tokenizer_chunk(lxb_html_tokenizer_t *tkz, const lxb_char_t *data,
 
     tkz->is_eof = false;
     tkz->status = LXB_STATUS_OK;
+    tkz->first = data;
     tkz->last = end;
 
     while (data < end) {
diff --git a/source/lexbor/html/tokenizer.h b/source/lexbor/html/tokenizer.h
index ba9602f..08d0d9a 100755
--- a/source/lexbor/html/tokenizer.h
+++ b/source/lexbor/html/tokenizer.h
@@ -72,6 +72,7 @@ struct lxb_html_tokenizer {
     lxb_char_t                       *pos;
     const lxb_char_t                 *end;
     const lxb_char_t                 *begin;
+    const lxb_char_t                 *first;
     const lxb_char_t                 *last;
 
     /* Entities */
diff --git a/source/lexbor/html/tree.c b/source/lexbor/html/tree.c
index 8c42990..28c97cc 100755
--- a/source/lexbor/html/tree.c
+++ b/source/lexbor/html/tree.c
@@ -484,6 +484,7 @@ lxb_html_tree_append_attributes(lxb_html_tree_t *tree,
 
         attr->node.local_name = token_attr->name->attr_id;
         attr->node.ns = ns;
+        attr->node.source_location = token_attr->name_begin - tree->tkz_ref->first;
 
         /* Fix for adjust MathML/SVG attributes */
         if (tree->before_append_attr != NULL) {
diff --git a/source/lexbor/html/tree.h b/source/lexbor/html/tree.h
index 231239d..bc0249e 100755
--- a/source/lexbor/html/tree.h
+++ b/source/lexbor/html/tree.h
@@ -266,8 +266,10 @@ lxb_inline lxb_dom_node_t *
 lxb_html_tree_create_node(lxb_html_tree_t *tree,
                           lxb_tag_id_t tag_id, lxb_ns_id_t ns)
 {
-    return (lxb_dom_node_t *) lxb_html_interface_create(tree->document,
+    lxb_dom_node_t *node = (lxb_dom_node_t *) lxb_html_interface_create(tree->document,
                                                         tag_id, ns);
+    node->source_location = (tree->tkz_ref->token->begin > tree->tkz_ref->first ? tree->tkz_ref->token->begin : tree->tkz_ref->begin) - tree->tkz_ref->first;
+    return node;
 }
 
 lxb_inline bool
